### DISCUSSION TITLE
TPI : File System : Duda ejemplos de uso de API

### DISCUSSION THREAD

**Discussion by jcastagno99** [dds-jv/foro#269] [Consultas de objetos] [answered]
Buenas! estoy teniendo un poco de problemas para implementar estos puntos en forma de test:

> Tenemos que leer de un archivo 3 campos: el primero de 4 bytes (C0), el segundo de 1 byte (C1), el tercero de 5 bytes (C2). Y escribir en otro archivo C0, un bloque 0x0, 0x10, 0x0, y luego C1 y C2.
Tenemos que leer un archivo completo, y escribirlo en otro archivo, en bloques de igual tamaño parametrizable.

1) A que se refiere con C0 , C1 y C2? Así nombra a los "campos" del archivo? y que serían esos "campos"?
2) Esto se puede hacer con una implementación guiada por la [resolución](https://docs.google.com/document/d/1l22DXR13J3XlcEkdwWsba5zg8gl_XwFA7cWf_LIOHHk/edit#) propuesta? porque según entiendo, en esta no se actualizan indices, se lee desde bufferStart hasta bufferEnd (en la operación de lectura como ejemplo), pero esos atributos son siempre el total del buffer y por lo que yo entiendo de lo primero que se pide una vez leído todos esos campos [**[2.1 : Puedo leer los 3 campos "de una" con un buffer de tamaño ideal [10 bytes] ? o es necesario hacerlo en varios casteos a lectura, cosa que esta implementación no soportaría a menos que use diferentes buffers de tamaños ideales a cada "campo"]**] debo escribír el primer campo, luego una serie de bloques y finalmente otros 2 campos, no se como hacer esto de lectura y escritura en intervalos con una solución similar a la propuesta, es decir, que no actualice punteros.

3) Esto no es una duda en sí, pero el segundo punto es mucho mas sencillo con esta implementación, porque siempre se lee y carga al array de bytes todo lo que el buffer aguante (y decir esto me hace pensar que esta mal) aunque lo que si no sabría bien es como obtener ese "tamaño del archivo" para hacer un buffer que lo tolere, podría "asumir" un tamaño muy grande pero no se hasta que punto es "penalizable"

Lo unico que se me ocurre es que debería refactorizar todo a mover indices (cosa que se me complicaría un poco y de poder ser preferiría evitar) pero por eso prefiero antes preguntar esto y ver si entiendo bien lo que se pide.

**Comment by jcastagno99** [comment, marked as answer]
También se me ocurre usar varios buffers, uno para cada campo, bufferC1(size de 4),  bufferC2(size de 1) y buffer C3(size de 5) y asumir que al leer el archivo (es decir, castear el read 3 veces sobre el file con los buffer c1 c2 y c3 en ese orden) el "puntero" se actualiza "por magia" para que me queden los buffers bien y ya castear la escritura como me piden, o leer todo de yapa (como dije antes) y separar el buffer en otros 3 para escribir luego cómodamente, mi problema acá es, cuanto puedo suponer? Hasta que punto puedo decir "esto lo hace el lowLevelFileSystem" (_magia_) o tengo que desarrollarlo yo.

A esto pude llegar pero no se si es la idea: 
![image](https://user-images.githubusercontent.com/49038828/120433396-8a994580-c351-11eb-8912-967ecdea967c.png)

**Reply by gastonprieto** [reply]
> A que se refiere con C0 , C1 y C2? Así nombra a los "campos" del archivo? y que serían esos "campos"?

Si

> Esto se puede hacer con una implementación guiada por la resolución propuesta? 

Si, te podes basar en eso.

> También se me ocurre usar varios buffers, uno para cada campo, bufferC1(size de 4), bufferC2(size de 1) y buffer C3(size de 5) y asumir que al leer el archivo (es decir, castear el read 3 veces sobre el file con los buffer c1 c2 y c3 en ese orden) el "puntero" se actualiza "por magia" para que me queden los buffers bien y ya castear la escritura como me piden, o leer todo de yapa 

No hay problema en asumir que el puntero del archivo avanza

**Reply by jcastagno99** [reply]
Buenisimo, gracias!
